### PR TITLE
vm/agent_exclusion

### PR DESCRIPTION
**Pull request by milanaghera-tenable**
This PR contains following change for Agent Exclusion API endpoint:

- change in folder structure as per v3 standard
- added search method and remove corresponding list method
- added marshmallow schema validation for validating the request body.
- test case will be found under the tests/io/v3/vm/agent_exclusion modules

### CODE REVIEW

**Review by parvdesai-tenable — APPROVED** [review]

File: tests/io/v3/vm/agent_exclusions/test_agent_exclusions.py [hunk 1]
```diff
@@ -0,0 +1,282 @@
+'''
+Test Case For Agent Exclusion
```
Comment (dhvanilshah-tenable) [inline tests/io/v3/vm/agent_exclusions/test_agent_exclusions.py:2]: Test Cases For Agent Exclusions APIs

File: tenable/io/v3/vm/__init__.py [hunk 2]
```diff
@@ -53,6 +55,14 @@ def agent_config(self):
         '''
         return AgentConfigAPI(self._api)
 
+    @property
+    def agent_exclusions(self):
+        '''
+        The interface object for the
+        :doc:`Agent Exclusions APIs <agent_exclusions>`.
```
[the remark this answers is not in the document]
Comment (milanaghera-tenable) [reply tenable/io/v3/vm/__init__.py:62]: Added as per the comment

File: tenable/io/v3/vm/agent_exclusions/api.py [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+from marshmallow import EXCLUDE
+from restfly.utils import dict_clean, dict_merge
+
+from tenable.io.v3.base.endpoints.explore import ExploreBaseEndpoint
+from tenable.io.v3.vm.agent_exclusions.schema import AgentExclusionSchema
+
+
+class AgentExclusionsAPI(ExploreBaseEndpoint):
+    '''
+    This will contain all methods related to agent exclusion
```
[the remark this answers is not in the document]
Comment (milanaghera-tenable) [reply tenable/io/v3/vm/agent_exclusions/api.py:27]: Updated.

File: tenable/io/v3/vm/agent_exclusions/api.py [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+            ...     timezone = 'Etc/UTC',
+            ...     weekdays=['mo', 'we', 'fr'],
+            ...     start_time=(
+            ...         datetime.utcnow()
+            ...     ).strftime('%Y-%m-%dT%H:%M:%SZ'),
+            ...     end_time=(
+            ...         datetime.utcnow() + timedelta(hours=1)
+            ...     ).strftime('%Y-%m-%dT%H:%M:%SZ')
+
+            Creating a monthly esxclusion:
```
[the remark this answers is not in the document]
Comment (milanaghera-tenable) [reply tenable/io/v3/vm/agent_exclusions/api.py:128]: updated.

File: tenable/io/v3/vm/agent_exclusions/schema.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    rrules = fields.Nested(RulesSchema, data_key='rrules', required=True)
+
+    @post_dump
+    def post_serialization(self, data, **kwargs):
+        if not data['enabled']:
+            data['starttime'] = (
+                datetime.utcnow()
+            ).strftime('%Y-%m-%dT%H:%M:%SZ')
+
+        if not data['enabled']:
```
[the remark this answers is not in the document]
Comment (milanaghera-tenable) [reply tenable/io/v3/vm/agent_exclusions/schema.py:64]: updated.

File: tests/io/v3/vm/agent_exclusions/test_agent_exclusions.py (hunk 1 above)
Comment (milanaghera-tenable) [reply tests/io/v3/vm/agent_exclusions/test_agent_exclusions.py:2]: updated.

File: tests/io/v3/vm/agent_exclusions/test_agent_exclusions.py [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+            'endtime': '2021-12-24T10:57:28Z',
+            'timezone': timezone,
+            'rrules': {
+                'freq': 'WEEKLY',
+                'interval': 1,
+                'byweekday': 'MO,WE,FR'
+            }
+        }
+    }
+    responses.add(
```
[the remark this answers is not in the document]
Comment (milanaghera-tenable) [reply tests/io/v3/vm/agent_exclusions/test_agent_exclusions.py:212]: added single line comment

File: tenable/io/v3/vm/agent_exclusions/schema.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
+'''
+Agent Exclusion API Endpoint Schemas
+'''
+from datetime import datetime, timedelta
+
+from marshmallow import Schema, fields, post_dump, validate
+from marshmallow.exceptions import ValidationError
+
+
+class RulesSchema(Schema):
```
[the remark this answers is not in the document]
Comment (milanaghera-tenable) [reply tenable/io/v3/vm/agent_exclusions/schema.py:10]: added docstring.

File: tenable/io/v3/vm/agent_exclusions/api.py [hunk 8]
```diff
@@ -0,0 +1,394 @@
...
+               weekdays: list = None
+               ) -> Dict:
+        '''
+        Creates a new agent exclusion.
+
+        :devportal:`agent-exclusions: create <agent-exclusions-create>`
+
+        Args:
+            agent_id (UUID): Than unique idetifier for agent
+            name (str): The name of the exclusion to create.
```
[the remark this answers is not in the document]
Comment (milanaghera-tenable) [reply tenable/io/v3/vm/agent_exclusions/api.py:52]: done.

File: tenable/io/v3/vm/agent_exclusions/api.py [hunk 9]
```diff
@@ -0,0 +1,394 @@
...
+            ...     agent_id = '00000000-0000-0000-0000-000000000000',
+            ...     exclusion_id = '00000000-0000-0000-0000-000000000000'
+            ... )
+        '''
+        return self._get(f'{agent_id}/exclusions/{exclusion_id}')
+
+    def edit(self,
+             agent_id: UUID,
+             exclusion_id: UUID,
+             day_of_month: int = None,
```
[the remark this answers is not in the document]
Comment (milanaghera-tenable) [reply tenable/io/v3/vm/agent_exclusions/api.py:256]: okay, done changes accordingly

File: tenable/io/v3/vm/agent_exclusions/api.py [hunk 10]
```diff
@@ -0,0 +1,394 @@
...
+            # Let's Merge rrules dict to payload
+            dict_merge(payload['schedule']['rrules'], rrules)
+
+        # Let's clean the payload
+        payload = dict_clean(payload)
+
+        # validate payload using marshmallow
+        schema = AgentExclusionSchema(
+            context={'valid_timezone': self._api._tz},
+            unknown=EXCLUDE
```
[the remark this answers is not in the document]
Comment (milanaghera-tenable) [reply tenable/io/v3/vm/agent_exclusions/api.py]: done.

File: tenable/io/v3/vm/agent_exclusions/api.py [hunk 11]
```diff
@@ -0,0 +1,389 @@
...
+    This class contains methods related to Agent Exclusions APIs
+    '''
+    _path: str = 'api/v3/agents'
+    _conv_json: bool = True
+
+    def create(self,
+               name: str,
+               start_time: str,
+               end_time: str = None,
+               timezone: str = 'Etc/UTC',
```
Comment (tomeara) [inline tenable/io/v3/vm/agent_exclusions/api.py:36]: Should this be 'ETC' instead? The casing looks a bit strange, and I'm wondering if it needs to be changed on the API side or not.

Comment (tomeara) [reply tenable/io/v3/vm/agent_exclusions/api.py:36]: Also, I would imagine that the default would be UTC alone. Is this even needed if we are providing start/end times in ISO8601?

Comment (milanaghera-tenable) [reply tenable/io/v3/vm/agent_exclusions/api.py:36]: @tomeara currently we are validating timezone using [scan/timezone](https://developer.tenable.com/reference/scans-timezones) API endpoint and I have checked the casing for Etc/UTC,  currently, we are using the same casing that we get from [scan/timezone](https://developer.tenable.com/reference/scans-timezones) API endpoint also the same casing is used in v2 as well.

Also now we are using the start/end time in iso8601 standard format right and so basically timestamp and timezone are the part of the start and end time. so we will remove the timezone parameter from the payload and incorporate these changes in API as well.

Please suggest your thoughts on this?

**Review by dhvanilshah-tenable — APPROVED** [review]

File: tenable/io/v3/vm/agent_exclusions/schema.py [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+
+        return data
+
+    @validates_schema
+    def validate_timezone(self, data, **kwargs):
+        if data['timezone'] not in self.context['timezones']:
+            raise ValueError('Invalid timezone field')
+
+
+class AgentExclusionSchema(Schema):
```
Comment (alpesh-te) [inline tenable/io/v3/vm/agent_exclusions/schema.py:90]: Correct the class name `AgentExclusionSchema` to `AgentExclusionsSchema` and make sure update related threads

File: tenable/io/v3/vm/agent_exclusions/schema.py [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+    '''
+    Schema for Agent Exclusion API
```
Comment (alpesh-te) [inline tenable/io/v3/vm/agent_exclusions/schema.py:92]: `Exclusion ` to `Exclusions `

File: tenable/io/v3/vm/agent_exclusions/api.py [hunk 14]
```diff
@@ -0,0 +1,387 @@
...
+    This class contains methods related to Agent Exclusions APIs
+    '''
+    _path: str = 'api/v3/agents/exclusions'
+    _conv_json: bool = True
+
+    def create(self,
+               name: str,
+               start_time: str,
+               end_time: Optional[str] = None,
+               timezone: Optional[str] = 'Etc/UTC',
```
Comment (alpesh-te) [inline tenable/io/v3/vm/agent_exclusions/api.py]: As per confluence comment timezone should be UTC https://confluence.eng.tenable.com/pages/viewpage.action?spaceKey=PORG&title=%5BSDK%5D+Agent+Exclusions

File: tests/io/v3/vm/agent_exclusions/test_agent_exclusions.py [hunk 15]
```diff
@@ -0,0 +1,364 @@
...
+    )
+
+    assert isinstance(res, dict)
+    assert res['name'] == name
+
+
+@responses.activate
+def test_delete(api):
+    '''
+    Test case for agent_exclusion delete method
```
Comment (alpesh-te) [inline tests/io/v3/vm/agent_exclusions/test_agent_exclusions.py:115]: Correct text `agent_exclusion` to `agent_exclusions ` need to update for many places

File: tests/io/v3/vm/agent_exclusions/test_agent_exclusions.py [hunk 16]
```diff
@@ -0,0 +1,364 @@
...
+
+    iterator = api.v3.vm.agent_exclusions.search(
+        fields=fields, filter=filter, sort=sort, limit=200
+    )
+    assert isinstance(iterator, SearchIterator)
+
+    exclusions_list = []
+    for item in iterator:
+        exclusions_list.append(item)
+    assert len(exclusions_list) == test_response['pagination']['total']
```
Comment (alpesh-te) [inline tests/io/v3/vm/agent_exclusions/test_agent_exclusions.py]: Get rid of loop.

**Review by alpesh-te — CHANGES_REQUESTED** [review]
Need to address minor changes

File: tests/io/v3/vm/agent_exclusions/test_agent_exclusions.py (hunk 16 above)
Comment (milanaghera-tenable) [reply tests/io/v3/vm/agent_exclusions/test_agent_exclusions.py]: Resolved.

File: tenable/io/v3/vm/agent_exclusions/api.py (hunk 14 above)
Comment (milanaghera-tenable) [reply tenable/io/v3/vm/agent_exclusions/api.py]: Resolved.

**Review by alpesh-te — CHANGES_REQUESTED** [review]
LGTM!

File: tests/io/v3/vm/agent_exclusions/test_agent_exclusions.py (hunk 15 above)
Comment (alpesh-te) [reply tests/io/v3/vm/agent_exclusions/test_agent_exclusions.py:115]: As discussed on last code, no need to change

File: tenable/io/v3/vm/agent_exclusions/schema.py (hunk 13 above)
Comment (alpesh-te) [reply tenable/io/v3/vm/agent_exclusions/schema.py:92]: As per last discussion no need to change

File: tenable/io/v3/vm/agent_exclusions/schema.py (hunk 12 above)
Comment (alpesh-te) [reply tenable/io/v3/vm/agent_exclusions/schema.py:90]: As per last discussion, no need to change

File: tenable/io/v3/vm/agent_exclusions/api.py (hunk 11 above)
Comment (alpesh-te) [reply tenable/io/v3/vm/agent_exclusions/api.py]: We will take care while IOS8601 while integration testing

**Review by alpesh-te — APPROVED** [review]

**Pull request merged**